packages/react-aria-components/src/Tree.tsx

```diff
@@ -0,0 +1,548 @@
+/*
+ * Copyright 2024 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {AriaTreeGridListProps, useTreeGridList, useTreeGridListItem} from '@react-aria/tree';
+import {BaseCollection, CollectionProps, CollectionRendererContext, ItemRenderProps, NodeValue, useCachedChildren, useCollection, useCollectionChildren, useShallowRender, useSSRCollectionNode} from './Collection';
+import {ButtonContext} from './Button';
+import {CheckboxContext} from './Checkbox';
+import {ContextValue, DEFAULT_SLOT, forwardRefType, Provider, RenderProps, ScrollableProps, SlotProps, StyleRenderProps, useContextProps, useRenderProps} from './utils';
+import {DisabledBehavior, Expandable, Key, LinkDOMProps} from '@react-types/shared';
+import {filterDOMProps, isAndroid, useObjectRef} from '@react-aria/utils';
+import {FocusScope,  mergeProps, useFocusRing, useGridListSelectionCheckbox, useHover, useLocalizedStringFormatter} from 'react-aria';
+import {Collection as ICollection, Node, SelectionBehavior, TreeState, useTreeState} from 'react-stately';
+// @ts-ignore
+import intlMessages from '../intl/*.json';
+import React, {createContext, ForwardedRef, forwardRef, HTMLAttributes, JSX, ReactElement, ReactNode, RefObject, useContext, useEffect, useMemo, useRef} from 'react';
+import {useControlledState} from '@react-stately/utils';
+
+class TreeCollection<T> implements ICollection<Node<T>> {
+  private flattenedRows: Node<T>[];
+  private keyMap: Map<Key, NodeValue<T>> = new Map();
+
+  constructor(opts) {
+    let {collection, expandedKeys} = opts;
+    let {flattenedRows, keyMap} = flattenTree<T>(collection, {expandedKeys});
+    this.flattenedRows = flattenedRows;
+    // Use generated keyMap because it contains the modified collection nodes (aka it adjusts the indexes so that they ignore the existence of the Content items)
+    this.keyMap = keyMap;
+  }
+
+  // TODO: should this collection's getters reflect the flattened structure or the original structure
+  // If we respresent the flattened structure, it is easier for the keyboard nav but harder to find all the nodes
+  *[Symbol.iterator]() {
+    yield* this.flattenedRows;
+  }
```

Comment:
At the moment this will give only the flattened structure (aka only visible rows), not sure what is most helpful/correct here when accessed by the hooks and stuff